### ISSUE TITLE
RFE: IDL should probably indicate whether method call can receive multiple replies

### ISSUE THREAD

**Issue by poettering** [varlink/varlink.github.io#26]
I think the IDL should have a way how we can mark method declarations in a way that indicates that it generates multiple replies (i.e. sets the "more" flag). This is a fundamental property of a method calls, and usually very specific ones, hence I think this really deserves some kind of visibility in the interface files.

My proposal: maybe a method call that receives multiple replies could be suffixed with a "+" to indicate this fact, i.e.:

```
method Foobar(x: bool) -> (y: int)+
```

would indicate a function that returns a series of integers, one in each reply.

This would imply that method replies are "similar", i.e. that the return type declaration sensibly describes *all* replies the same. Which I think makes sense to expect.

**Comment by poettering** [comment]
or hmm, varlink idl so far preferred prefix notation for arrays and such, i.e. it's `[]int`, and hence maybe we should also do the "+" as prefix?

```
method Foobar(x: bool) -> +(y:int)
```

or so?

anyway, ideas?

**Comment by poettering** [comment]
hmm, and whether an upgrade is possible should probably be shown in a similar way. maybe:

```c
method Foobar(x: bool) -> &(y:int)
```

or is all that too cryptic? ideas?

**Comment by poettering** [comment]
It might actually make sense to indicate two distinct concepts:

1. Whether a method *requires* the "more" flag to be set (i.e. not setting the flag would be an error)
2. Whether a method *supports* the "more" flag to be set (function can be called with the flag set or not, both are fine, and make sense)

**Comment by dolmen** [comment]
I agree that the IDL should indicate whether method call can receive multiple replies.

See also #41 and #42 about `oneway` and `upgrade`.

**Comment by zeenix** [comment]
> maybe a method call that receives multiple replies could be suffixed with a "+"

Wouldn't this be a breaking change? The existing varlink interface parsers will have to be modified to handle this, no?

**Comment by poettering** [comment]
yes

**Comment by emersion** [comment]
FWIW, I see all methods as being callable with the `more` flag, but most might answer with just a single final reply.

**Comment by zeenix** [comment]
> FWIW, I see all methods as being callable with the `more` flag, but most might answer with just a single final reply.

Perhaps in theory but I learnt the hard way that in practice, some methods require `more` to be set as they use this as a way to split an otherwise very big message into smaller chunks.

**Comment by emersion** [comment]
That sounds like a pretty weird API design, perhaps a workaround for the lack of streaming JSON decoding support? Not sure Varlink was designed to build APIs like this.

Can you give an example?

**Comment by zeenix** [comment]
> That sounds like a pretty weird API design, perhaps a workaround for the lack of streaming JSON decoding support? Not sure Varlink was designed to build APIs like this.

Yeah, I agree but I think it would be very difficult to convince people to not do this, especially since Varlink doesn't provide any other means to accomplish this.

> Can you give an example?

https://github.com/systemd/systemd/issues/38301

**Comment by poettering** [comment]
> That sounds like a pretty weird API design, perhaps a workaround for the lack of streaming JSON decoding support? Not sure Varlink was designed to build APIs like this.
> 
> Can you give an example?

The way I think of varlink "more" is that it's about subscription and enumeration (which are very closely related concepts). If you set "more" then you set things up that you want to have multiple replies, because there are multiple objects, and that can be either immediately/now (as in enumeration), or can come at a later time (as in subscription).

Hence, I think it's hightly relevant if the IDL tells you if this is one of these interfaces or not.

**Comment by bachorp** [comment]
> It might actually make sense to indicate two distinct concepts:
> 
> 1. Whether a method *requires* the "more" flag to be set (i.e. not setting the flag would be an error)
> 2. Whether a method *supports* the "more" flag to be set (function can be called with the flag set or not, both are fine, and make sense)

If we have 1., then something like 2. could be achieved without further help from the IDL. For example,

```nim
# This method either
# - gives one response for each thing (if `more` is used)
# - gives only the first thing (otherwise)
method ListThings () -> (thing: object)

# This method either
# - periodically reports the progress on doing something until it's done (if `more` is used)
# - returns `{}` after doing something (otherwise)
method DoSomething () -> (progress: ?float)
```

could alternatively be implemented as follows (using some new syntax for methods with possibly multiple responses).

```nim
# `head: true` tells to give only the first thing
method ListThings (head: ?bool) -> ...(thing: object)

# will respond when something has been done
method DoSomething () -> ()
# will periodically report the progress on doing something until it's done
method DoSomethingWithProgress () -> ...(progress: float)
```

In fact I believe it makes sense to avoid *overloading* methods using `more`.

**Comment by Snaipe** [comment]
FWIW, the varlink.org website uses this example in the interface definition page:

```
method Foo(a: int, b: MyType) -> (bar: MyOtherType, baz: float, more: (i: int, f: float, s: string))
```

I don't think consensus has been reached on how to express that a method call returns multiple replies, yet the choice of wording with `more` made me think that this is how extra replies would be encoded, _especially_ since the other args followed a foo/bar/baz sequence. It seemed to suggest that this is how the IDL encoded the meaning of having more replies. It's only by searching through issues in the varlink github org that I happened to stumble upon this page.

Reading through this issue, it seems odd that this example isn't actually something that was brought up. Code generators could simply interpret a `more` output parameter to be the format of any further replies, with the rest of the output parameters being the format of the final reply. One advantage is that this does not require any change to the IDL syntax.

At the very least, until consensus is reached, I'd strongly recommend putting the lack of representation for more/upgrade in the IDL in the F.A.Q, and change the above example to rename "more" to some inane "foo" or "quxx" to avoid any further confusion.

**Comment by dolmen** [comment]
@Snaipe This issue is about the `more` which is not in the output parameters but in the `Call` data: https://varlink.org/Method-Call

**Comment by Snaipe** [comment]
> [@Snaipe](https://github.com/Snaipe) This issue is about the `more` which is not in the output parameters but in the `Call` data: https://varlink.org/Method-Call

I understand this well; I'm just pointing out that the example I cited in my comment can be misinterpreted as being the output parameter that gets sent if call.more is set to true.

In other words, I interpreted `method Foo(a: int, b: MyType) -> (bar: MyOtherType, baz: float, more: (i: int, f: float, s: string))` as meaning: "_method Foo takes parameters `(a: int, b: MyType)`, and if called with `more` set to `true`, then it returns one or more structs of the form `(i: int, f: float, s: string)`, followed by one final reply of the form `(bar: MyOtherType, baz: float)`. Otherwise, if `more` is set to `false`, return one reply of the form `(bar: MyOtherType, baz: float)`_".

The reason why I interpreted it as such was because of the choice of naming; why name the output parameter "more" instead of "quxx" when the other two follow a foo-bar-baz sequence?

I perfectly understand that there is no way to represent call.more=true being supported by a method in the IDL, my argument is that the example should be amended to `method Foo(a: int, b: MyType) -> (bar: MyOtherType, baz: float, quxx: (i: int, f: float, s: string))` to stop the confusion from spreading.

The secondary point to my argument is that _this construct above_ could well be a way to express in the IDL that a method call supports more, and without any syntactic change to the parser. Said differently, taking the first example, `method Foobar(x: bool) -> (y: int)+` could be expressed as `method Foobar(x: bool) -> (y: int, more: (y: int))`, _by convention_. The obvious drawback is that this prevents "more" from ever being meaningfully used as output parameter, but this may be less burdensome than introducing new syntax for this.
